File: SegmentTree/minimumInSubArray/minInSubArr.cpp
#include<iostream>
#include<algorithm>
#include<climits>

void buildSegTree(int* arr, int* segTree, int start, int end, int treeIndex) {
    
    // Base case
    if(start == end) {
        segTree[treeIndex] = arr[start];
        return;
    }
    
    // Calculate Mid
    int mid {(start+end) / 2};

    // Call recursion to build left subtree
    buildSegTree(arr,segTree,start,mid,2*treeIndex);

    // Call recursion to build right subtree
    buildSegTree(arr,segTree, mid+1,end,2*treeIndex+1);
    segTree[treeIndex] = std::min(segTree[2*treeIndex], segTree[2*treeIndex + 1]);
}

int query(int* segTree,int start, int end, int treeNode, int left, int right) {
    // complete outside
    if(start > right || end < left) return INT_MAX;

    // complete inside
    if(start >= left && end <= right) return segTree[treeNode];

    // partially
    int mid {(start + end) / 2};
    int ans1 {query(segTree,start,mid,2*treeNode,left,right)};
    int ans2 {query(segTree,mid+1,end,2*treeNode+1,left,right)};
    return std::min(ans1,ans2);
}

void updateTree(int* arr, int* segTree, int start, int end, int treeIndex, int index, int newValue) {
    if (start == end) {
        segTree[treeIndex] = newValue;
        arr[index] = newValue;
        return;
    }

    int mid {(start+end) / 2};

    if(index > mid) {
        updateTree(arr,segTree,mid+1,end,2*treeIndex+1,index,newValue);
    } else {
        updateTree(arr,segTree,start,mid,2*treeIndex,index,newValue);
    }

    // Updating tree node
    segTree[treeIndex] = std::min(segTree[2*treeIndex], segTree[2*treeIndex + 1]);
}

int main(){
    int n,q;
    std::cin >> n >> q;
    int arr[n];
    for(int i{}; i<n; i++) std::cin >> arr[i];
    
    int* segTree {new int[4*n]};
    for(int i{}; i<n; i++) segTree[i] = 0;
    buildSegTree(arr,segTree,0,n-1,1);


    while(q--) {
        char c;
        std::cin >> c;
        int left,right;
        std::cin >> left >> right;
        if(c == 'q') {
            int ans {query(segTree,0,n-1,1,left-1,right-1)};
            std::cout << ans << std::endl;
        } else {
            updateTree(arr,segTree,0,n-1,1,left-1,right);
        }
    }
}

File: SegmentTree/minimumInSubArray/minInSubArr.js
(function(){


    function buildTree(arr, segTree, start, end, treeNode) {
        if(start === end) {
            segTree[treeNode] = arr[start];
            return;
        }

        let mid = parseInt((start + end) / 2);
        buildTree(arr,segTree,start,mid, 2*treeNode);
        buildTree(arr,segTree,mid+1,end,2*treeNode+1);
        segTree[treeNode] = Math.min(segTree[2*treeNode], segTree[2*treeNode+1]);
    }

    function update(arr,segTree,start,end,treeNode,index, newValue) {
        if (start === end) {
            segTree[treeNode] = newValue;
            arr[index] = newValue;
            return;
        }

        let mid = parseInt((start+mid) / 2);
        if(index > mid) update(arr,segTree,mid+1,end,2*treeNode+1,index,newValue);
        else update(arr,segTree,start,mid,2*treeNode,index,newValue);
        
        segTree[treeNode] = Math.min(segTree[2*treeNode], segTree[2*treeNode+1]);
    }

    function query(segTree,start,end,treeNode,left,right) {
        if(start > right || end < left) return Number.MAX_SAFE_INTEGER;

        if(start >= left && end <= right) return segTree[treeNode];

        let mid = parseInt((start + end) / 2);
        let ans1 = query(segTree,start,mid,2*treeNode,left,right);
        let ans2 = query(segTree,mid+1,end,2*treeNode+1,left,right);
        return Math.min(ans1,ans2);
    }

    (function main(){
        
        let arr = [1,5,2,4,3];
        let n = 5;

        let segTree = [];
        buildTree(arr,segTree,0,n-1,1);

        let ans = query(segTree,0,n-1,1,0,3);
        console.log(ans);
        console.log(query(segTree,0,n-1,1,2,4));
    }());


}());